File: problems/114. Flatten Binary Tree to Linked List/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def flatten(self, root: TreeNode) -> None:
        if not root:
            return
        curr = root
        stack = []
        if curr.right:
            stack.append(curr.right)
        if curr.left:
            stack.append(curr.left)

        while stack:
            nextNode = stack.pop()
            curr.left = None
            curr.right = nextNode

            if nextNode.right:
                stack.append(nextNode.right)
            if nextNode.left:
                stack.append(nextNode.left)

            curr = nextNode

File: problems/114. Flatten Binary Tree to Linked List/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    void flatten(TreeNode *root)
    {
        if (root == nullptr)
            return;
        stack<TreeNode *> stack;
        TreeNode *curr = root;
        if (curr->right)
            stack.push(curr->right);
        if (curr->left)
            stack.push(curr->left);

        while (stack.size())
        {
            TreeNode *node = stack.top();
            stack.pop();
            curr->left = nullptr;
            curr->right = node;
            if (node->right)
                stack.push(node->right);
            if (node->left)
                stack.push(node->left);
            curr = node;
        }
    }
};